### PR TITLE
Routing

### PR DESCRIPTION
**Pull request by JuliusKarl**
Major Additions

- Styling error pages (400/404/500)
- Bypassing '__typename' error by redirecting to `error/500`

### CODE REVIEW

**Comment by rosemcc** [comment]
Can you please update @uoa/error-pages to the latest version? (v0.1.1) .. hopefully it doesn't break anything :)

**Comment by rosemcc** [comment]
The routing seems kind of buggy.. e.g. say if I navigate from /subhub/research-impact to /research-impact/planning-for-impact, it works fine. But then if I refresh the page it goes back to home. Also if I just paste in the url for /research-impact/planning-for-impact into the browser it can't find the route.

**Comment by rosemcc** [comment]
routing.ts: Should it redirect to error:404 instead of home if no route match is found?
e.g.
{
    path: '**',
    redirectTo: 'error/404',
    pathMatch: 'full'
  }

**Comment by rosemcc** [comment]
The error styling looks great 👍

File: research-hub-web/src/app/components/articles/articles.component.html [hunk 1]
```diff
@@ -1,226 +1,89 @@
...
+                        : ''">
+                <div class="banner-content">
+
+                    <!-- Title -->
+                    <h1 role="heading" aria-level="1" class="content-title" ngClass.lt-md="content-title-mobile"
+                        [style]="article.banner ? 'color: white; text-shadow: 0px 0px 8px #333333;' : ''">{{ article.title }}</h1>
+
+                    <!-- Summary -->
+                    <p class="content-summary" ngClass.sm="content-summary-sm" ngClass.xs="content-summary-xs"
+                        [style]="article.banner ? 'color: white; text-shadow: 0px 0px 8px #333333;' : ''">
```
Comment (Hganavak) [inline research-hub-web/src/app/components/articles/articles.component.html:46]: Lots of inline styles going on here

File: research-hub-web/src/app/components/error-routing/error-routing.module.ts [hunk 2]
```diff
@@ -1,12 +1,35 @@
...
+            
+        for (let [key, value] of Object.entries(this.ErrorPageContent)) {
+            value.content += `
+                <br><br>
+                <a href="/">
+                <div role="button" class="standard-button">
+                    <p>ResearchHub Home</p>
+                </div>
+                </a>`;
+        }
```
Comment (Hganavak) [inline research-hub-web/src/app/components/error-routing/error-routing.module.ts:33]: Not sure if I love or hate this... 🤷  Very JSX-esque

File: research-hub-web/src/app/components/error-routing/error-routing.scss [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+@import '../../../typography.scss';
+
+.content-wrapper {
+    text-align: center!important;
+    h1 {
+        color: $dark-blue!important;
+        font-weight: bold!important;
+    }
+    a div {
+        p {
```
Comment (Hganavak) [inline research-hub-web/src/app/components/error-routing/error-routing.scss:11]: We have paragraphs inside of `divs` inside of links...? 😩

File: research-hub-web/src/app/components/events/events.component.html [hunk 4]
```diff
@@ -1,119 +1,77 @@
...
 
     <ng-template #specificationsTable>
         <div 
             class="specifications section">
-            <table class="specifications-table">
-                <tr>
-                <th>Details</th>
-                <th>Description</th>
+            <table class="specifications-table" role="table">
+                <tr role="row">
```
Comment (Hganavak) [inline research-hub-web/src/app/components/events/events.component.html:23]: aria-nice 😻

File: research-hub-web/src/app/components/home/browse/browse.component.scss [hunk 5]
```diff
@@ -1 +1,50 @@
-@import "../../../../styles";
\ No newline at end of file
+@import '../../../../styles.scss';
+@import '../../../../themes.scss';
+@import '../../../../typography.scss';
```
Comment (Hganavak) [inline research-hub-web/src/app/components/home/browse/browse.component.scss:3]: What happened to import aliases?

Comment (JuliusKarl) [reply research-hub-web/src/app/components/home/browse/browse.component.scss:3]: oh yeah jest was disabling it for some reason but it should be fine to use import alias again

File: research-hub-web/src/app/components/home/home.component.scss [hunk 6]
```diff
@@ -1,2 +1,58 @@
...
+    @extend .margin-bottom;
+    padding: 16px 0;
+    width: 95vw;
+    margin: 0 auto;
+    text-align: center;
+}
+
+
+.grey-background {
+    padding-bottom: 40px;
```
Comment (Hganavak) [inline research-hub-web/src/app/components/home/home.component.scss:20]: This and `.teal-background` seem like misleading classnames if they're adding margins... and kinda unnecessary TBH.

File: research-hub-web/src/app/components/home/research-activity/research-activity.component.scss [hunk 7]
```diff
@@ -1,31 +1,123 @@
...
+}
+
 
-  &.mobile-view {
-    height: $ra-chip-height-touch;
+/**
+ * Research Activity Container
+ */
+.research-activity-content {
+  & .outer-panel {
```
Comment (Hganavak) [inline research-hub-web/src/app/components/home/research-activity/research-activity.component.scss:23]: Is the space here intentional? (Between `&` and `.`)

File: research-hub-web/src/app/components/home/research-activity/research-activity.component.scss [hunk 8]
```diff
@@ -1,31 +1,123 @@
...
+      justify-content: center;
+      background: map-get($hub-teal, 500);
+      border: 4px solid map-get($hub-teal, 500);
+      transition: background 0.11s ease-in-out, color 0.1s ease-in-out;
+      @include size(large) {
+        width: 55%;
+        margin-left: -5vw;
+        transform: skew(30deg);
+      }
+      & * {
```
Comment (Hganavak) [inline research-hub-web/src/app/components/home/research-activity/research-activity.component.scss:63]: Again here, all of these selectors are redundant.

```scss
.hub {
    .sam { color: #ccc }
}

```
Equals:

```html
<element class="hub">
  …
    <element class="sam">
```
---
And:
```scss
.hub {
    &.sam { color: #ccc }
}
```

Equals: 

```html
<element class="hub sam">
```
---
But what you're doing here:
```scss
.hub {
    & .sam { color: #ccc }
}
```

Equals: 

```html
<element class="hub">
  …
    <element class="sam">
```

I.e. the same thing as the first option, the `&` is not needed. That's the whole point of nesting classes in Sass.

File: research-hub-web/src/app/components/search-page/search-page.component.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    this.searchBarService.createResultsList();
+    this.initialPages();
+  }
+
+  // Create the initial page lsit
+  public async initialPages() {
+
+    // Updating results when searched
+    this.resultSub$ = this.searchBarService.resultsChange.subscribe(data => {
+      this.allCurrentPages = data.map((x) => {return { ...x };});
```
Comment (Hganavak) [inline research-hub-web/src/app/components/search-page/search-page.component.ts:50]:
```typescript
this.allCurrentPages = data.map((x) => {return { ...x };});
this.allCurrentPages = data.map(x => { ...x });
```
Personal pref but looks tidier to me.

File: research-hub-web/src/app/components/error-routing/error-routing.module.ts (hunk 2 above)
Comment (rosemcc) [reply research-hub-web/src/app/components/error-routing/error-routing.module.ts:33]: @Hganavak Didn't we tell you we're going to migrate the Hub to react?

File: research-hub-web/cypress/integration/search.e2e.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    it('displays search content', () => {
+        cy.get('input').type('{enter}');
+        cy.contains('Results found');
+    });
+
+    it('displays number of results found', () => {
+        cy.get('input').type('ffffffffff');
+        cy.get('input').type('{enter}');
+        cy.contains('0 Results found');
```
Comment (rosemcc) [inline research-hub-web/cypress/integration/search.e2e.js:24]: Maybe don't specify a number (0) here, to make the test a bit more robust

File: research-hub-web/src/app/components/search-page/search-page.component.html [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+                        <mat-option value="">Default</mat-option>
+                        <mat-option value="A-Z">Alphabetical</mat-option>
+                    </div>
+                </mat-select>
+            </div>
+        </div>
+
+        <!-- No results message -->
+        <div *ngIf="searchBarService.getTotalPages() == 0" class="no-results">
+            <p>Sorry, your search for "<i><b>{{ searchBarService.getSearchText() }}</b></i>" in All Categories, did not match anything on the ResearchHub.</p>
```
Comment (rosemcc) [inline research-hub-web/src/app/components/search-page/search-page.component.html:115]: remove 'in All Categories'?

**Comment by rosemcc** [comment]
Nice Julius, tests passing so should be good to merge

**Pull request merged**